controllers/kafkamanagement/kafkauser_controller.go

```diff
@@ -79,291 +75,318 @@ func (r *KafkaUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+					l.Error(err, "Cannot update Kafka user secret",
+						"secret name", secret.Name,
+						"secret namespace", secret.Namespace)
+					r.EventRecorder.Eventf(user, models.Warning, models.UpdatedEvent,
+						"Cannot assign Kafka user to a k8s secret. Reason: %v", err)
+
+					return models.ReconcileRequeue, nil
+				}
+			}
+
+			continue
 		}
-		return r.handleUpdateKafkaUser(ctx, user, l), nil
-	default:
-		l.Info("Unhandled event", "annotations", user.Annotations[models.ResourceStateAnnotation])
-		return models.ExitReconcile, nil
-	}
-}
 
-func (r *KafkaUserReconciler) handleCreateKafkaUser(
-	ctx context.Context,
-	user *v1beta1.KafkaUser,
-	l logr.Logger,
-) reconcile.Result {
-	if user.Status.ID == "" {
-		l.Info(
-			"Creating Kafka User resource",
-			"kafka cluster ID", user.Spec.ClusterID,
-			"initial permissions", user.Spec.InitialPermissions,
-			"kafka user options", user.Spec.Options,
-		)
+		if clusterEvent == models.DeletingEvent {
+			userID := user.GetKafkaUserID(clusterID, username)
+			err = r.API.DeleteKafkaUser(userID, instaclustr.KafkaUserEndpoint)
+			if err != nil {
+				l.Error(err, "cannot delete Kafka User resource",
+					"initial permissions", user.Spec.InitialPermissions,
+					"kafka user options", user.Spec.Options,
+					"kafka user metadata", user.ObjectMeta,
+				)
+				r.EventRecorder.Eventf(
+					user, models.Warning, models.DeletionFailed,
+					"Resource deletion on the Instaclustr is failed. Reason: %v",
+					err,
+				)
+
+				return models.ReconcileRequeue, nil
+			}
+			r.EventRecorder.Eventf(
+				user, models.Normal, models.DeletionStarted,
+				"Resource deletion request is sent to the Instaclustr API.",
+			)
+			patch = user.NewPatch()
+			delete(user.Status.ClustersEvents, clusterID)
+			err = r.Status().Patch(ctx, user, patch)
+			if err != nil {
+				l.Error(err, "cannot patch Kafka User resource status",
+					"initial permissions", user.Spec.InitialPermissions,
+					"kafka user options", user.Spec.Options,
+					"kafka user metadata", user.ObjectMeta,
+				)
+				r.EventRecorder.Eventf(
+					user, models.Warning, models.PatchFailed,
+					"Resource status patch is failed. Reason: %v",
+					err,
+				)
+
+				return models.ReconcileRequeue, nil
+			}
 
-		iKafkaUser := user.Spec.ToInstAPI()
-		username, password, err := r.getKafkaUserCredsFromSecret(user.Spec)
-		if err != nil {
-			l.Error(
-				err, "Cannot get Kafka User creds from secret",
-				"kafka user spec", iKafkaUser,
+			l.Info("Kafka User has been deleted",
+				"initial permissions", user.Spec.InitialPermissions,
+				"kafka user options", user.Spec.Options,
 			)
+
 			r.EventRecorder.Eventf(
-				user, models.Warning, models.FetchFailed,
-				"Fetch user credentials from secret is failed. Reason: %v",
-				err,
+				user, models.Normal, models.Deleted,
+				"Resource is deleted",
 			)
-			return models.ReconcileRequeue
+
+			controllerutil.RemoveFinalizer(user, models.DeletionUserFinalizer+clusterID)
```

Comment:
Use the `GetDeletionUserFinalizer()` method instead please

Reply:
done